js/modules/k6/experimental/streams/readable_stream_default_controller.go

```diff
@@ -0,0 +1,485 @@
+// the stream being canceled, used to clean up the state stored in the
+// controller and inform the underlying source.
+//
+// It implements the ReadableStreamDefaultControllerCancelSteps [specification]
+// algorithm.
+//
+// [specification]: https://streams.spec.whatwg.org/#readable-stream-default-controller-cancel-steps
+func (controller *ReadableStreamDefaultController) cancelSteps(reason any) *goja.Promise {
+	// 1. Perform ! ResetQueue(this).
+	controller.resetQueue()
+
+	// 2. Let result be the result of performing this.[[cancelAlgorithm]], passing reason.
+	result := controller.cancelAlgorithm(reason)
+
+	// 3. Perform ! ReadableStreamDefaultControllerClearAlgorithms(this).
+	controller.clearAlgorithms()
+
+	// 4. Return result.
+	if p, ok := result.Export().(*goja.Promise); ok {
+		return p
+	}
+
+	return newRejectedPromise(controller.stream.vu, newError(RuntimeError, "cancel algorithm error"))
+}
+
+// pullSteps performs the controller’s steps that run when a default reader
+// is read from, used to pull from the controller any queued chunks, or
+// pull from the underlying source to get more chunks.
+//
+// It implements the [ReadableStreamDefaultControllerPullSteps] specification
+// algorithm.
+//
+// [ReadableStreamDefaultControllerPullSteps]: https://streams.spec.whatwg.org/#rs-default-controller-private-pull
+func (controller *ReadableStreamDefaultController) pullSteps(readRequest ReadRequest) {
+	// 1. Let stream be this.[[stream]].
+	stream := controller.stream
+
+	// 2. If this.[[queue]] is not empty,
+	if controller.queue.Len() > 0 {
+		// 2.1. Let chunk be ! DequeueValue(this).
+		chunk, err := controller.queue.Dequeue()
+		if err != nil {
+			common.Throw(stream.vu.Runtime(), err)
+		}
+
+		// 2.2. If this.[[closeRequested]] is true and this.[[queue]] is empty,
+		if controller.closeRequested && controller.queue.Len() == 0 {
+			// 2.2.1. Perform ! ReadableStreamDefaultControllerClearAlgorithms(this).
+			controller.clearAlgorithms()
+			// 2.2.2. Perform ! ReadableStreamClose(stream).
+			stream.close()
+		} else {
+			// 2.3. Otherwise, perform ! ReadableStreamDefaultControllerCallPullIfNeeded(this).
+			controller.callPullIfNeeded()
+		}
+
+		// 2. 4. Perform readRequest’s chunk steps, given chunk.
+		readRequest.chunkSteps(chunk)
+	} else { // 3. Otherwise,
+		// 3.1. Perform ! ReadableStreamAddReadRequest(stream, readRequest).
+		stream.addReadRequest(readRequest)
+
+		// 3.2. Perform ! ReadableStreamDefaultControllerCallPullIfNeeded(this).
+		controller.callPullIfNeeded()
+	}
+}
+
+// releaseSteps performs the controller’s steps that run when a reader is
+// released, used to clean up reader-specific resources stored in the controller.
+//
+// It implements the ReadableStreamDefaultControllerReleaseSteps [specification]
+// algorithm.
+//
+// [specification]: https://streams.spec.whatwg.org/#readable-stream-default-controller-release-steps
+func (controller *ReadableStreamDefaultController) releaseSteps() {
```

Comment:
it is a bit strange that this is empty, and the link about doesn't work for me :thinking:

Reply:
Good catch! Fixing the link, the correct one: https://streams.spec.whatwg.org/#readable-stream-default-controller-release-steps

Turns out, the specification contains a single instruction which is to return in the case of the default controller. I'm modifying the code to reflect that.

The specification mentions that this is to implement the [[ReleaseSteps]] contract, and the ByteStream implementation which we haven't implemented yet has much more steps for it: https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamcontroller-releasesteps